server/src/main/java/org/opensearch/repositories/blobstore/BlobStoreRepository.java

```diff
@@ -988,13 +1048,50 @@ private void asyncCleanupUnlinkedShardLevelBlobs(
         if (filesToDelete != null) {
             threadPool.executor(ThreadPool.Names.SNAPSHOT).execute(ActionRunnable.wrap(listener, l -> {
                 try {
-                    deleteFromContainer(blobContainer(), filesToDelete);
+                    if (remoteStoreLockManagerFactory != null) {
+                        for (String fileToDelete : filesToDelete) {
+                            if (fileToDelete.contains(SHALLOW_SNAPSHOT_PREFIX)) {
+                                String[] fileToDeletePath = fileToDelete.split("/");
+                                String indexId = fileToDeletePath[1];
+                                String shardId = fileToDeletePath[2];
+                                String shallowSnapBlob = fileToDeletePath[3];
+                                String snapshotUUID = shallowSnapBlob.substring(
+                                    SHALLOW_SNAPSHOT_PREFIX.length(),
+                                    shallowSnapBlob.length() - ".dat".length()
+                                );
+                                BlobContainer shardContainer = blobStore().blobContainer(indicesPath().add(indexId).add(shardId));
+                                RemoteStoreShardShallowCopySnapshot remoteStoreShardShallowCopySnapshot =
+                                    REMOTE_STORE_SHARD_SHALLOW_COPY_SNAPSHOT_FORMAT.read(
+                                        shardContainer,
+                                        snapshotUUID,
+                                        namedXContentRegistry
+                                    );
+                                String indexUUID = remoteStoreShardShallowCopySnapshot.getIndexUUID();
+                                String remoteStoreRepoForIndex = remoteStoreShardShallowCopySnapshot.getRemoteStoreRepository();
+                                // Releasing lock file before deleting the shallow-snap-UUID file because in case of any failure while
+                                // releasing the lock file, we would still have the shallow-snap-UUID file and that would be used during
+                                // next delete operation for releasing this lock file
+                                RemoteStoreMetadataLockManager remoteStoreMetadataLockManager = remoteStoreLockManagerFactory
+                                    .newLockManager(remoteStoreRepoForIndex, indexUUID, shardId);
+                                remoteStoreMetadataLockManager.release(
+                                    FileLockInfo.getLockInfoBuilder().withAcquirerId(snapshotUUID).build()
+                                );
+                            }
+                            // Deleting the shard blob
+                            blobContainer().deleteBlobsIgnoringIfNotExists(Arrays.asList(fileToDelete));
+                        }
+                    } else {
+                        deleteFromContainer(blobContainer(), filesToDelete);
```

Comment:
this logic can be simplified to release the lock files if the `remoteStoreLockManagerFactory` is not null. Then after releasing the lock files delete all the blobs in bulk using `deleteFromContainer(blobContainer(), filesToDelete)`